tls/s2n_cipher_suites.c

```diff
@@ -1079,10 +1079,19 @@ struct s2n_cipher_suite *s2n_cipher_suite_from_wire(const uint8_t cipher_suite[S
 int s2n_set_cipher_as_client(struct s2n_connection *conn, uint8_t wire[S2N_TLS_CIPHER_SUITE_LEN])
 {
     notnull_check(conn);
+    notnull_check(conn->secure.cipher_suite);
+    struct s2n_cipher_suite *cipher_suite;
 
     /* See if the cipher is one we support */
-    conn->secure.cipher_suite = s2n_cipher_suite_from_wire(wire);
-    S2N_ERROR_IF(conn->secure.cipher_suite == NULL, S2N_ERR_CIPHER_NOT_SUPPORTED);
+    cipher_suite = s2n_cipher_suite_from_wire(wire);
+    ENSURE_POSIX(cipher_suite != NULL, S2N_ERR_CIPHER_NOT_SUPPORTED);
+
+    /* Verify cipher suite sent in server hello is the same as sent in hello retry */
+    if (s2n_conn_get_current_message_type(conn) == SERVER_HELLO && s2n_is_hello_retry_handshake(conn)) {
+        ENSURE_POSIX(conn->secure.cipher_suite->iana_value == cipher_suite->iana_value, S2N_ERR_CIPHER_TYPE);
```

Comment:
The error message for S2N_ERR_CIPHER_TYPE is "Unknown cipher type used". I'm not sure that fits here. You might want "S2N_ERR_CIPHER_NOT_SUPPORTED" bc they're trying to use a cipher we didn't offer or "S2N_ERR_BAD_KEY_SHARE" because that's how it would show up [if they choose a cipher suite the client didn't send](https://github.com/awslabs/s2n/blob/master/tls/extensions/s2n_server_key_share.c#L108-L109).